### ISSUE TITLE
Mock of HTTP requests is too rigid

### ISSUE THREAD

**Issue by guilhermeluizsp** [ava-innersource/Liquid-Application-Framework-1.0-deprecated#102]
Currently there's only one way to mock HTTP requests (when using `AbstractApiWrapper` or `LightAPI`): We must add a file "stubAPIData.json" with the scheme below, inside the "Mock" directory

```json
{
  "Hosts": [
    {
      "Name": "[the full path to the LightAPI config]",
      "Methods": [
        {
          "Name": "[the HttpMethod]",
          "Route": "[the request url]",
          "Response": {}
        }
      ]
    }
  ]
}
```

https://github.com/Avanade/Liquid-Application-Framework/blob/1d6c37c44fb5bbb6f84ba2f1d1a8a1b3d646c2ad/src/Liquid.Domain/API/AbstractAPIWrapper.cs#L257-L288

so, if we wanted to mock a GET request to "www.avanade.com/endpoint", this file would look like this:

```json
{
  "Hosts": [
    {
      "Name": "LightApi:Avanade",
      "Methods": [
        {
          "Name": "GET",
          "Route": "http://www.avanade.com/endpoint",
          "Response": {
            "test": "value"
          }
        }
      ]
    }
  ]
}
```

This approach of mocking http requests has a couple of issues:

### It mixes test code with production code
I'm really into "**You shouldn't mix test code with production code**", and while this is a personal opinion, I still think it should be taken into consideration (issues with mocks shouldn't raise framework updates).

### It can only return DomainResponse
This approach can only return `DomainResponse`s. This is problematic since third party APIs certainly don't always return responses matching the `DomainResponse` structure.

https://github.com/Avanade/Liquid-Application-Framework/blob/1d6c37c44fb5bbb6f84ba2f1d1a8a1b3d646c2ad/src/Liquid.Domain/API/AbstractAPIWrapper.cs#L271-L277

### It forces the 200 - OK status
One benefit of mocking an HTTP request is to be able to simulate lots of scenarios. Some of them don't always return 200. What if we want to assert that something will be executed only when the api returns 404?

https://github.com/Avanade/Liquid-Application-Framework/blob/1d6c37c44fb5bbb6f84ba2f1d1a8a1b3d646c2ad/src/Liquid.Domain/API/AbstractAPIWrapper.cs#L272-L274

### It doesn't take the request body into consideration
Same thing as above. What if we want to assert that something is executed only when we send a specific field with a certain value?

## Summing up
There are plenty of libraries made specifically for tests out there ([Moq](https://github.com/moq/moq) is my personal favorite). Those are well tested libraries with tons of features. Let's embrace them and not pollute our code with unit test conditions.

I know Liquid is supposed to be simple yet effective, but is this coupling with test code a worthy effort? We can still keep our unit tests simple by taking these kind of code out of the source code. Should we start an open discussion for this?

**Comment by bruno-brant** [comment]
While this is certainly an enhancement, I'm marking it priority; it may even be more important than tests.

My take here is simple - we should allow traditional mocking. This means some things:

- Current implementations must be changed, because they are coupled to LightApi (if they, unfortunately, followed Liquid's guide). 

- We must allow for objects to dynamic resolve the api implementation. The most obvious solution is a factory, but I have some other ideas I'd like to propose. 

- The service locator is still a kind of pain since we can't mock the dependency directly, but we should be able to mock it through registering a singleton instance. I'm still, however, trying to cleanly solve it for threading scenarios, since xunit tests run in parallel.
